### ISSUE TITLE
flatMapLatest + combine + first causes 

### ISSUE THREAD

**Issue by ansman** [Kotlin/kotlinx.coroutines#4034]
**Describe the bug**

Firstly, it's possible this is expected behavior but it certainly doesn't seem that way.

We have a flow that we collect using `first()`. This flow is a combination of multiple upstream flows but they all emit their items synchronously. The downstream flow limits the number of emissions to a single element then uses `awaitCompletion()` to avoid terminating the flow. This is the flow we call `first()` on.

Under some circumstances, we are seeing the `first()` call resuming _after_ the scope has been cancelled which is causing a crash for us.

If either the `flatMapLatest` or `combine` is removed, the code behaves as expected, although `first()` never resumes at these `x

I know there are no real guarantees that code won't be running while canceled. But there are few things that makes me think that this doesn't apply in this case:
- There is no non-cooperative suspension as far as I know.
- Only a single thread is involved.

**Provide a Reproducer**
This reproduces it on an Android device, but it's possible it can reproed on the JVM too.
```kotlin
// This must be run on the main thread
val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)
println("Before launch")
scope.launch {
    flowOf(listOf(1, 2, 3))
        .flatMapLatest { numbers ->
            combine(numbers.map { flowOf(it) }) { it.sum() }
        }
        .onEach { println("Got item") }
        .first()
    println("After first, scope is active: ${scope.isActive}")
}
println("After launch")
scope.cancel()
```
I expect this to print:
```
Before launch
Got item
After first, scope is active: true
After launch
```

Instead I'm seeing this:
```
Before launch
Got item
After launch
After first, scope is active: false
```

**Comment by dkhalanskyjb** [comment]
Could not reproduce with coroutines 1.7.3 using Android instrumented tests. The output for me is:
```
Before launch
Got item
After launch
```
Due to cancellation, the "After first" `println` didn't execute.

**Comment by ansman** [comment]
Yes, that's right. It does _not_ repro in an instrumented test for us either. But it does when you put that code in `Activity.onCreate` it does. I'll attach a complete project.

**Comment by ansman** [comment]
Here is a complete example. Run it and see the output in logcat.
[Coroutinesbug.zip](https://github.com/Kotlin/kotlinx.coroutines/files/14195972/Coroutinesbug.zip)

**Comment by dkhalanskyjb** [comment]
Ok, reproduced.

**Comment by ansman** [comment]
@dkhalanskyjb Can you remove the "waiting for clarification" label so it doesn't get auto closed?

**Comment by jxdabc** [comment]
The test case below could reproduce this issue.

```kotlin
val job = launch(start = UNDISPATCHED) {
    flow {
        coroutineScope {
            emitAll(produce(start = UNDISPATCHED) {
                channel.send(Unit)
            })
        }
    }.first()
    expectUnreached()
}
job.cancel()
```

That is what happened：

1. `job` launched ---> `first()` ran `flow` ---> `produce` launched and `send(Unit)` to `channel` (two launches were `UNDISPATCHED` / `immediate`) 
2. `send(Unit)` suspended because `channel` had no buffer, and `emitAll()` ran. `emitAll()` took the `Unit` from `channel`, emitted it to `first()` and threw AbortFlowException.
3. coroutineScope suspended waiting for the subtask `produce` to complete, and `job.cancel()` ran. `coroutineScope` was in exception state and the `CancellationException` was overridden by the previous `AbortFlowException`.
4. `first()` catched the  `AbortFlowException` (inside `collectWhile`) and failed to response to `cancel()`.

This PR should fix this issue https://github.com/jxdabc/kotlinx.coroutines/commit/9bb7d70eb7a68686c1b4eec3b465fdace35b12a5

**Comment by dkhalanskyjb** [comment]
Thanks for the analysis! This doesn't explain the difference between Android and instrumented tests, but indeed, the reproducer is valid. It can be simplified further:
```kotlin
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.*

fun main() {
    runBlocking {
        val job = launch(start = CoroutineStart.UNDISPATCHED) {
            flow {
                try {
                    println(1)
                    emit(Unit)
                } finally {
                    runCatching { yield() }
                }
            }.first()
            println(3)
            error("unreached")
        }
        println(2)
        job.cancel()
    }
}
```

**Comment by jxdabc** [comment]
> Thanks for the analysis! This doesn't explain the difference between Android and instrumented tests.

If ExampleInstrumentedTest is like something below. That won't reproduce.

This is because `first()` and `cancel()` run in different threads concurrently. (main thread and test thread).


```kotlin
val dis = Dispatchers.Main.immediate
class ExampleInstrumentedTest {
    @Test
    fun useAppContext() {
        val dis = Dispatchers.Main.immediate
        val scope = CoroutineScope(dis)
        println("Before launch")
        scope.launch {
            flowOf(listOf(1, 2, 3))
                .flatMapLatest { numbers ->
                    combine(numbers.map { flowOf(it) }) { it.sum() }
                }
                .first()
            println("After first, scope is active: ${scope.isActive}")
        }
        println("After launch")
        scope.cancel()
    }
}
```
